# Language level evaluation

In [29]:
from notebook_init import notebook_init, notebook_imports

notebook_init()
notebook_imports()

from src.bots.evaluation_bot import LanguageLevelEvaluationBot
 
from sklearn.metrics import accuracy_score

%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [30]:
examples = [
    ('The author\'s nuanced use of symbolism and intricate narrative structure elevate the novel, making it a captivating exploration of the human psyche', 'A1', 1),
    ('I recently read an interesting article about the impact of technology on modern society.', 'A1', 2),
    ('I like to read. My favourite book is Alice in Wonderland', 'A1', 3),
    ('The author\'s nuanced use of symbolism and intricate narrative structure elevate the novel, making it a captivating exploration of the human psyche', 'A2', 1),
    ('I would have really enjoyed the boat ride, such a pitty it is cancelled.', 'A2', 2),
    ('I enjoy reading books and watching movies in my free time.', 'A2', 3),
    ('The author\'s nuanced use of symbolism and intricate narrative structure elevate the novel, making it a captivating exploration of the human psyche','B1', 1),
    ('The author\'s eloquent use of language captivates readers and imbues the narrative with a profound sense of emotional depth.', 'B1', 2),
    ('I like to read books and watch movies in my free time.', 'B1', 3),
    ('I love my mom a lot.', 'B2', 1),
    ('I like to read books and watch movies in my free time.', 'B2', 2),
    ('I recently read an interesting article about climate change and its effects on marine life.', 'B2', 3),
    ('I like to read a lot.', 'C1', 1),
    ('The recent increase in online shopping has led to a significant shift in consumer behavior and preferences.', 'C1', 2),
    ('The author\'s eloquent use of language captivates readers and imbues the narrative with a profound sense of emotional depth.', 'C1', 3),
    ('I like to read books and watch movies.', 'C2', 1),
    ('I recently read an interesting article about climate change and its effects on marine life.', 'C2', 2),
    ('His eloquent discourse on the complexities of quantum physics captivated even the most astute minds in the audience.', 'C2', 3)
]

correct_ratings = [e[2] for e in examples] 

In [6]:
def count_diff(l1, l2):
    return sum( 1 if abs(e[0] - e[1]) > 1 else 0 for e in zip(l1,l2) )

### Alternative 0

In [5]:
bot = LanguageLevelEvaluationBot(0)

ratings = [ bot.evaluate('English', m[1], m[0]) for m in examples]

acc =  accuracy_score(ratings, correct_ratings)   
diff_g_e = count_diff(ratings, correct_ratings)

print('Acc: ', acc)
print('Num diff > 1: ', diff_g_e )
print(ratings)
print(correct_ratings)

Acc:  0.5
Num diff > 1:  0
[1, 1, 2, 1, 2, 3, 1, 1, 3, 2, 3, 3, 2, 3, 3, 2, 3, 3]
[1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3]


In [ ]:
sum_acc = 0
sum_num_diff = 0
n_runs = 10

bot = LanguageLevelEvaluationBot(0)

for i in range(n_runs):
    

    ratings = [ bot.evaluate('English', m[1], m[0]) for m in examples]

    sum_acc +=  accuracy_score(ratings, correct_ratings)   
    sum_num_diff += count_diff(ratings, correct_ratings)/len(examples)
    
print('Average acc: ', sum_acc/n_runs)
print('Average percentage diff > 1: ', sum_num_diff/n_runs )

### Alternative 1

In [ ]:
bot = LanguageLevelEvaluationBot(1)

ratings = [ bot.evaluate('English', m[1], m[0]) for m in examples]

acc =  accuracy_score(ratings, correct_ratings)   
diff_g_e = count_diff(ratings, correct_ratings)


print('Acc: ', acc)
print('Num diff > 1: ', diff_g_e )
print(ratings)
print(correct_ratings)

In [36]:
sum_acc = 0
sum_num_diff = 0
n_runs = 10

bot = LanguageLevelEvaluationBot(1)

for i in range(n_runs):
    

    ratings = [ bot.evaluate('English', m[1], m[0]) for m in examples]

    sum_acc +=  accuracy_score(ratings, correct_ratings)   
    sum_num_diff += count_diff(ratings, correct_ratings)/len(examples)
    
print('Average acc: ', sum_acc/n_runs)
print('Average percentage diff > 1: ', sum_num_diff/n_runs )

Average acc:  0.5555555555555555
Average percentage diff > 1:  0.05


### Alternative 2

In [12]:
bot = LanguageLevelEvaluationBot(2)

ratings = [ bot.evaluate('English', m[1], m[0], print_prompt=True) for m in examples]

acc =  accuracy_score(ratings, correct_ratings)   
diff_g_e = count_diff(ratings, correct_ratings)


print('Acc: ', acc)
print('Num diff > 1: ', diff_g_e )
print(ratings)
print(correct_ratings)

Acc:  0.4444444444444444
Num diff > 1:  0
[1, 1, 3, 1, 2, 2, 2, 1, 3, 2, 3, 3, 2, 3, 3, 2, 3, 3]
[1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3]


In [47]:
sum_acc = 0
sum_num_diff = 0
n_runs = 10

bot = LanguageLevelEvaluationBot(2)

for i in range(n_runs):
    

    ratings = [ bot.evaluate('English', m[1], m[0]) for m in examples]

    sum_acc +=  accuracy_score(ratings, correct_ratings)   
    sum_num_diff += count_diff(ratings, correct_ratings)/len(examples)
    
print('Average acc: ', sum_acc/n_runs)
print('Average percentage diff > 1: ', sum_num_diff/n_runs )

Average acc:  0.5222222222222223
Average percentage diff > 1:  0.09444444444444447


### Alternative 3

In [ ]:
bot = LanguageLevelEvaluationBot(3)

ratings = [ bot.evaluate('English', m[1], m[0], print_prompt=True) for m in examples]

acc =  accuracy_score(ratings, correct_ratings)   
diff_g_e = count_diff(ratings, correct_ratings)


print('Acc: ', acc)
print('Num diff > 1: ', diff_g_e )
print(ratings)
print(correct_ratings)

In [40]:
sum_acc = 0
sum_num_diff = 0
n_runs = 10

bot = LanguageLevelEvaluationBot(3)

for i in range(n_runs):
    

    ratings = [ bot.evaluate('English', m[1], m[0]) for m in examples]

    sum_acc +=  accuracy_score(ratings, correct_ratings)   
    sum_num_diff += count_diff(ratings, correct_ratings)/len(examples)
    
print('Average acc: ', sum_acc/n_runs)
print('Average percentage diff > 1: ', sum_num_diff/n_runs )

Average acc:  0.5277777777777779
Average percentage diff > 1:  0.01111111111111111


### Alternative 4

In [ ]:
bot = LanguageLevelEvaluationBot(prompt_alternative=4, level_description_fname='../../templates/language_levels_cefr_extended.json')

ratings = [ bot.evaluate('English', m[1], m[0], print_prompt=True) for m in examples]

acc =  accuracy_score(ratings, correct_ratings)   
diff_g_e = count_diff(ratings, correct_ratings)


print('Acc: ', acc)
print('Num diff > 1: ', diff_g_e )
print(ratings)
print(correct_ratings)

In [41]:
sum_acc = 0
sum_num_diff = 0
n_runs = 10

bot = LanguageLevelEvaluationBot(4)

for i in range(n_runs):
    

    ratings = [ bot.evaluate('English', m[1], m[0]) for m in examples]

    sum_acc +=  accuracy_score(ratings, correct_ratings)   
    sum_num_diff += count_diff(ratings, correct_ratings)/len(examples)
    
print('Average acc: ', sum_acc/n_runs)
print('Average percentage diff > 1: ', sum_num_diff/n_runs )

Average acc:  0.5166666666666667
Average percentage diff > 1:  0.005555555555555555


### Alternative 5

In [ ]:
bot = LanguageLevelEvaluationBot(prompt_alternative=5)

ratings = [ bot.evaluate('English', m[1], m[0], print_prompt=True) for m in examples]

acc =  accuracy_score(ratings, correct_ratings)   
diff_g_e = count_diff(ratings, correct_ratings)


print('Acc: ', acc)
print('Num diff > 1: ', diff_g_e )
print(ratings)
print(correct_ratings)

In [43]:
sum_acc = 0
sum_num_diff = 0
n_runs = 10

bot = LanguageLevelEvaluationBot(5)

for i in range(n_runs):
    

    ratings = [ bot.evaluate('English', m[1], m[0]) for m in examples]

    sum_acc +=  accuracy_score(ratings, correct_ratings)   

    sum_num_diff += count_diff(ratings, correct_ratings)/len(examples)
    
print('Average acc: ', sum_acc/n_runs)
print('Average percentage diff > 1: ', sum_num_diff/n_runs )

Average acc:  0.5111111111111111
Average percentage diff > 1:  0.0


### Alternative 6

In [ ]:
bot = LanguageLevelEvaluationBot(6, level_description_fname='../../templates/language_levels_cefr.json')

ratings = [ bot.evaluate('English', m[1], m[0], print_prompt=True) for m in examples]

acc =  accuracy_score(ratings, correct_ratings)   
diff_g_e = count_diff(ratings, correct_ratings)


print('Acc: ', acc)
print('Num diff > 1: ', diff_g_e )
print(ratings)
print(correct_ratings)

In [46]:
sum_acc = 0
sum_num_diff = 0
n_runs = 10

bot = LanguageLevelEvaluationBot(6)

for i in range(n_runs):
    

    ratings = [ bot.evaluate('English', m[1], m[0]) for m in examples]

    sum_acc +=  accuracy_score(ratings, correct_ratings)   

    sum_num_diff += count_diff(ratings, correct_ratings)/len(examples)
    
print('Average acc: ', sum_acc/n_runs)
print('Average percentage diff > 1: ', sum_num_diff/n_runs )

Average acc:  0.5722222222222222
Average percentage diff > 1:  0.05


|Prompt alternative | Average accuracy | Average percentage of difference > 1 |
|-------------------|------------------|--------------------------------------|
| 0 | 0.522 | 0.0 |
| 1 |  0.556 | 0.05 |
| 2 | 0.522| 0.09 |
| 3 | 0.528 |0.011|
| 4 | 0.516 |0.001 |
| 5 | 0.511 | 0.0 |
| 6 | 0.572| 0.05|